File: android/src/main/java/com/mcodex/nitroarchive/HybridArchiveModule.kt
package com.mcodex.nitroarchive

import android.content.Context
import com.margelo.nitro.archive.*
import com.margelo.nitro.core.ArrayBuffer
import com.margelo.nitro.core.Promise
import com.mcodex.nitroarchive.application.CreateArchiveUseCase
import com.mcodex.nitroarchive.application.OpenArchiveUseCase
import com.mcodex.nitroarchive.domain.ArchiveDomainError
import com.mcodex.nitroarchive.infrastructure.memory.BufferArchiveInput
import com.mcodex.nitroarchive.infrastructure.storage.LocalFileInput
import com.mcodex.nitroarchive.infrastructure.zip.Zip4jEngine
import com.mcodex.nitroarchive.platform.AndroidArchiveAccessInspector
import java.io.File
import java.io.RandomAccessFile

class HybridArchiveModule(private val context: Context) : HybridArchiveModuleSpec() {
    private val engine = Zip4jEngine()
    private val accessInspector = AndroidArchiveAccessInspector(context)

    override fun getCapabilities(): NativeArchiveCapabilities {
        return NativeArchiveCapabilities(
            platform = NativePlatform.ANDROID,
            readableFormats = arrayOf("zip"),
            writableFormats = arrayOf("zip"),
            compressionMethods = arrayOf("store", "deflate"),
            encryptionMethods = arrayOf("none", "zip-crypto", "aes-128", "aes-256"),
            supportsFilePaths = true,
            supportsInputUris = true,
            supportsOutputUris = true,
            supportsDirectoryUris = true,
            supportsAtomicPathWrites = false,
            supportsSecurityScopedUrls = false,
            supportsZip64 = true
        )
    }

    override fun checkFileAccess(path: String, mode: String): Promise<NativeAccessOutcome> = Promise.async {
        val accessMode = if (mode == "write" || mode == "create-children") {
            com.mcodex.nitroarchive.application.ports.ArchiveAccessMode.WRITE
        } else {
            com.mcodex.nitroarchive.application.ports.ArchiveAccessMode.READ
        }
        val report = accessInspector.checkFileAccess(path, accessMode)
        NativeAccessOutcome(
            ok = report.accessible,
            accessible = report.accessible,
            readable = report.readable,
            writable = report.writable,
            persistent = true,
            securityScoped = false,
            providerBacked = false,
            seekable = true,
            reason = report.reason,
            recoveryAction = if (report.accessible) "none" else "choose-document",
            error = if (report.accessible) null else NativeArchiveFailure(ArchiveErrorCode.E_PERMISSION_DENIED, report.reason ?: "Access denied", null, null, null, null, null)
        )
    }

    override fun checkUriAccess(uri: String, mode: String): Promise<NativeAccessOutcome> = Promise.async {
        val accessMode = if (mode == "write" || mode == "create-children") {
            com.mcodex.nitroarchive.application.ports.ArchiveAccessMode.WRITE
        } else {
            com.mcodex.nitroarchive.application.ports.ArchiveAccessMode.READ
        }
        val report = accessInspector.checkUriAccess(uri, accessMode)
        NativeAccessOutcome(
            ok = report.accessible,
            accessible = report.accessible,
            readable = report.readable,
            writable = report.writable,
            persistent = false,
            securityScoped = false,
            providerBacked = true,
            seekable = null,
            reason = report.reason,
            recoveryAction = if (report.accessible) "none" else "reselect-document",
            error = if (report.accessible) null else NativeArchiveFailure(ArchiveErrorCode.E_PERMISSION_DENIED, report.reason ?: "Access denied", null, null, null, null, null)
        )
    }

    override fun detectFile(path: String): Promise<NativeDetectionOutcome> = Promise.async {
        val file = File(path)
        if (!file.exists() || file.length() < 4) {
            return@async NativeDetectionOutcome(ok = true, format = null, confidence = 0.0, extensionMatches = null, error = null)
        }
        val raf = RandomAccessFile(file, "r")
        val signature = ByteArray(4)
        raf.readFully(signature)
        raf.close()
        val isZip = signature[0] == 0x50.toByte() && signature[1] == 0x4B.toByte() && signature[2] == 0x03.toByte() && signature[3] == 0x04.toByte()
        val extMatch = path.endsWith(".zip", ignoreCase = true)
        NativeDetectionOutcome(
            ok = true,
            format = if (isZip) "zip" else null,
            confidence = if (isZip) 1.0 else 0.0,
            extensionMatches = extMatch,
            error = null
        )
    }

    override fun detectUri(uri: String): Promise<NativeDetectionOutcome> = Promise.async {
        val androidUri = android.net.Uri.parse(uri)
        if (androidUri.scheme == "file") {
            detectFile(androidUri.path ?: uri).await()
        } else {
            try {
                val inputStream = context.contentResolver.openInputStream(androidUri)
                val signature = inputStream?.use { it.readNBytes(4) }
                if (signature == null || signature.size < 4) {
                    NativeDetectionOutcome(ok = true, format = null, confidence = 0.0, extensionMatches = null, error = null)
                } else {
                    val isZip = signature[0] == 0x50.toByte() && signature[1] == 0x4B.toByte() && signature[2] == 0x03.toByte() && signature[3] == 0x04.toByte()
                    NativeDetectionOutcome(ok = true, format = if (isZip) "zip" else null, confidence = if (isZip) 1.0 else 0.0, extensionMatches = null, error = null)
                }
            } catch (e: Exception) {
                NativeDetectionOutcome(ok = false, format = null, confidence = 0.0, extensionMatches = null, error = NativeArchiveFailure(ArchiveErrorCode.E_IO, e.message ?: "Failed to read URI", null, null, null, null, null))
            }
        }
    }

    override fun detectBuffer(data: ArrayBuffer): Promise<NativeDetectionOutcome> = Promise.async {
        val bytes = data.toByteArray()
        if (bytes.size < 4) {
            return@async NativeDetectionOutcome(ok = true, format = null, confidence = 0.0, extensionMatches = null, error = null)
        }
        val isZip = bytes[0] == 0x50.toByte() && bytes[1] == 0x4B.toByte() && bytes[2] == 0x03.toByte() && bytes[3] == 0x04.toByte()
        NativeDetectionOutcome(
            ok = true,
            format = if (isZip) "zip" else null,
            confidence = if (isZip) 1.0 else 0.0,
            extensionMatches = null,
            error = null
        )
    }

    override fun openFile(path: String, options: NativePathOpenOptions): Promise<NativeOpenOutcome> = Promise.async {
        try {
            val input = LocalFileInput(path)
            val useCase = OpenArchiveUseCase(engine)
            val result = useCase.execute(input)
            val reader = HybridArchiveReader(
                inspection = result.inspection,
                session = result.session,
                engine = engine
            )
            NativeOpenOutcome(ok = true, reader = reader, error = null)
        } catch (e: ArchiveDomainError) {
            NativeOpenOutcome(ok = false, reader = null, error = NativeArchiveFailure(
                code = mapCode(e.code),
                message = e.message ?: e.code,
                operationId = null, entryPath = null, source = path, destination = null, nativeCode = null
            ))
        } catch (e: Exception) {
            NativeOpenOutcome(ok = false, reader = null, error = NativeArchiveFailure(
                code = ArchiveErrorCode.E_IO,
                message = e.message ?: "I/O error",
                operationId = null, entryPath = null, source = path, destination = null, nativeCode = null
            ))
        }
    }

    override fun openUri(uri: String, options: NativeUriOpenOptions): Promise<NativeOpenOutcome> = Promise.async {
        val androidUri = android.net.Uri.parse(uri)
        if (androidUri.scheme == "file") {
            val pathOptions = NativePathOpenOptions(
                password = options.password,
                maxEntriesToIndex = options.maxEntriesToIndex,
                maxCentralDirectoryBytes = options.maxCentralDirectoryBytes
            )
            return@async openFile(androidUri.path ?: uri, pathOptions).await()
        }

        try {
            val ctx = context
            val inputStream = ctx.contentResolver.openInputStream(androidUri)
                ?: throw ArchiveDomainError.FileNotAvailable("Cannot open input stream for URI: $uri")
            val bytes = inputStream.use { it.readBytes() }
            val bufferInput = BufferArchiveInput(bytes)
            val useCase = OpenArchiveUseCase(engine)
            val result = useCase.execute(bufferInput)
            val reader = HybridArchiveReader(
                inspection = result.inspection,
                session = result.session,
                engine = engine
            )
            NativeOpenOutcome(ok = true, reader = reader, error = null)
        } catch (e: ArchiveDomainError) {
            NativeOpenOutcome(ok = false, reader = null, error = NativeArchiveFailure(
                code = mapCode(e.code), message = e.message ?: e.code,
                operationId = null, entryPath = null, source = uri, destination = null, nativeCode = null
            ))
        } catch (e: Exception) {
            NativeOpenOutcome(ok = false, reader = null, error = NativeArchiveFailure(
                code = ArchiveErrorCode.E_IO,
                message = e.message ?: "I/O error reading content:// URI",
                operationId = null, entryPath = null, source = uri, destination = null, nativeCode = null
            ))
        }
    }

    override fun openBuffer(data: ArrayBuffer, options: NativeBufferOpenOptions): Promise<NativeOpenOutcome> = Promise.async {
        try {
            val bytes = data.toByteArray()
            val input = BufferArchiveInput(bytes)
            val useCase = OpenArchiveUseCase(engine)
            val result = useCase.execute(input)
            val reader = HybridArchiveReader(
                inspection = result.inspection,
                session = result.session,
                engine = engine
            )
            NativeOpenOutcome(ok = true, reader = reader, error = null)
        } catch (e: ArchiveDomainError) {
            NativeOpenOutcome(ok = false, reader = null, error = NativeArchiveFailure(
                code = mapCode(e.code), message = e.message ?: e.code,
                operationId = null, entryPath = null, source = options.name, destination = null, nativeCode = null
            ))
        } catch (e: Exception) {
            NativeOpenOutcome(ok = false, reader = null, error = NativeArchiveFailure(
                code = ArchiveErrorCode.E_IO, message = e.message ?: "I/O error",
                operationId = null, entryPath = null, source = options.name, destination = null, nativeCode = null
            ))
        }
    }

    override fun create(request: NativeCreationRequest): HybridNativeCreationTaskSpec {
        return HybridCreationTask(engine = engine, request = request)
    }

    private fun mapCode(code: String): ArchiveErrorCode {
        return try {
            ArchiveErrorCode.valueOf(code)
        } catch (_: IllegalArgumentException) {
            ArchiveErrorCode.E_INTERNAL
        }
    }
}

File: ios/HybridArchiveModule.swift
import Foundation
import NitroModules

final class HybridArchiveModule: HybridArchiveModuleSpec {
  private let engine: ArchiveEngine
  private let accessInspector: ArchiveAccessInspecting

  override init() {
    self.engine = SSZipArchiveEngine()
    self.accessInspector = IOSArchiveAccessInspector()
    super.init()
  }

  func getCapabilities() throws -> NativeArchiveCapabilities {
    NativeArchiveCapabilities(
      platform: .ios,
      readableFormats: ["zip"],
      writableFormats: ["zip"],
      compressionMethods: ["store", "deflate"],
      encryptionMethods: ["none", "zip-crypto", "aes-256"],
      supportsFilePaths: true,
      supportsInputUris: true,
      supportsOutputUris: false,
      supportsDirectoryUris: false,
      supportsAtomicPathWrites: false,
      supportsSecurityScopedUrls: true,
      supportsZip64: true
    )
  }

  func checkFileAccess(path: String, mode: String) throws -> Promise<NativeAccessOutcome> {
    Promise.async {
      let accessMode: ArchiveAccessMode = mode == "write" || mode == "create-children" ? .write : .read
      let report = self.accessInspector.checkFileAccess(path: path, mode: accessMode)
      return NativeAccessOutcome(
        ok: report.accessible,
        accessible: report.accessible,
        readable: report.readable,
        writable: report.writable,
        persistent: true,
        securityScoped: false,
        providerBacked: false,
        seekable: true,
        reason: report.reason,
        recoveryAction: report.accessible ? "none" : "choose-document",
        error: report.accessible ? nil : NativeArchiveFailure(code: .ePermissionDenied, message: report.reason ?? "Access denied", operationId: nil, entryPath: nil, source: nil, destination: nil, nativeCode: nil)
      )
    }
  }

  func checkUriAccess(uri: String, mode: String) throws -> Promise<NativeAccessOutcome> {
    Promise.async {
      let accessMode: ArchiveAccessMode = mode == "write" || mode == "create-children" ? .write : .read
      let report = self.accessInspector.checkUriAccess(uri: uri, mode: accessMode)
      return NativeAccessOutcome(
        ok: report.accessible,
        accessible: report.accessible,
        readable: report.readable,
        writable: report.writable,
        persistent: false,
        securityScoped: true,
        providerBacked: false,
        seekable: true,
        reason: report.reason,
        recoveryAction: report.accessible ? "none" : "reselect-document",
        error: report.accessible ? nil : NativeArchiveFailure(code: .ePermissionDenied, message: report.reason ?? "Access denied", operationId: nil, entryPath: nil, source: nil, destination: nil, nativeCode: nil)
      )
    }
  }

  func detectFile(path: String) throws -> Promise<NativeDetectionOutcome> {
    Promise.async {
      guard let signature = try? Data(contentsOf: URL(fileURLWithPath: path), options: .alwaysMapped),
            signature.count >= 4 else {
        return NativeDetectionOutcome(ok: true, format: nil, confidence: 0, extensionMatches: nil, error: nil)
      }
      let isZip = signature[0] == 0x50 && signature[1] == 0x4B && signature[2] == 0x03 && signature[3] == 0x04
      let extMatch = path.hasSuffix(".zip")
      return NativeDetectionOutcome(
        ok: true,
        format: isZip ? "zip" : nil,
        confidence: isZip ? 1.0 : 0,
        extensionMatches: extMatch,
        error: nil
      )
    }
  }

  func detectUri(uri: String) throws -> Promise<NativeDetectionOutcome> {
    Promise.async {
      guard let url = URL(string: uri) else {
        return NativeDetectionOutcome(ok: false, format: nil, confidence: 0, extensionMatches: nil, error: NativeArchiveFailure(code: .eInvalidArgument, message: "Invalid URI", operationId: nil, entryPath: nil, source: nil, destination: nil, nativeCode: nil))
      }
      return try await self.detectFile(path: url.path).await()
    }
  }

  func detectBuffer(data: ArrayBuffer) throws -> Promise<NativeDetectionOutcome> {
    Promise.async {
      guard data.size >= 4 else {
        return NativeDetectionOutcome(ok: true, format: nil, confidence: 0, extensionMatches: nil, error: nil)
      }
      let bytes = Data(bytes: data.data, count: data.size)
      let isZip = bytes[0] == 0x50 && bytes[1] == 0x4B && bytes[2] == 0x03 && bytes[3] == 0x04
      return NativeDetectionOutcome(ok: true, format: isZip ? "zip" : nil, confidence: isZip ? 1.0 : 0, extensionMatches: nil, error: nil)
    }
  }

  func openFile(path: String, options: NativePathOpenOptions) throws -> Promise<NativeOpenOutcome> {
    Promise.async {
      await self.openInput(LocalFileInput(path: path))
    }
  }

  func openUri(uri: String, options: NativeUriOpenOptions) throws -> Promise<NativeOpenOutcome> {
    Promise.async {
      guard let url = URL(string: uri), url.isFileURL else {
        return NativeOpenOutcome(ok: false, reader: nil, error: NativeArchiveFailure(code: .eInvalidArgument, message: "Not a file URL", operationId: nil, entryPath: nil, source: nil, destination: nil, nativeCode: nil))
      }
      return await self.openInput(LocalFileInput(path: url.path))
    }
  }

  func openBuffer(data: ArrayBuffer, options: NativeBufferOpenOptions) throws -> Promise<NativeOpenOutcome> {
    Promise.async {
      await self.openInput(BufferArchiveInput(data: Data(bytes: data.data, count: data.size)))
    }
  }

  private func openInput(_ input: ArchiveInput) async -> NativeOpenOutcome {
    let useCase = OpenArchiveUseCase(engine: self.engine)
    do {
      let result = try await useCase.execute(input: input)
      return NativeOpenOutcome(ok: true, reader: HybridArchiveReader(
        inspection: result.inspection,
        session: result.session,
        engine: self.engine
      ), error: nil)
    } catch let error as ArchiveDomainError {
      return NativeOpenOutcome(ok: false, reader: nil, error: NativeArchiveFailure(code: self.mapCode(error.code), message: "\(error)", operationId: nil, entryPath: nil, source: nil, destination: nil, nativeCode: nil))
    } catch {
      return NativeOpenOutcome(ok: false, reader: nil, error: NativeArchiveFailure(code: .eIo, message: error.localizedDescription, operationId: nil, entryPath: nil, source: nil, destination: nil, nativeCode: nil))
    }
  }

  func create(request: NativeCreationRequest) throws -> (any HybridNativeCreationTaskSpec) {
    HybridCreationTask(engine: engine, request: request)
  }

  private func mapCode(_ code: String) -> ArchiveErrorCode {
    ArchiveErrorCode(fromString: code) ?? .eInternal
  }
}
